### PR TITLE
Add A Command To Unmatch User And Job

### PR DESCRIPTION
**Pull request by AlekKwek18**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        boolean hasJobMatchedContact = jobToUnmatch.hasMatched(contactIdentifier);
+
+        if (!hasContactMatchedJob && !hasJobMatchedContact) {
+            throw new CommandException(MESSAGE_ALREADY_UNMATCHED);
+        }
+
+        assert hasContactMatchedJob == hasJobMatchedContact;
+        // Unmatch the contact from the job by removing the associations
+        Person unmatchedContact = unmatchContactFromJob(contactToUnmatch);
+        Job unmatchedJob = unmatchJobFromContact(jobToUnmatch, contactIdentifier);
```
Comment (KengHian) [inline src/main/java/seedu/address/logic/commands/UnmatchCommand.java:89]: The naming of these functions seems a bit unclear, for the first line, if you are trying to remove the reference of a job in a contact and return said contact, shouldn't the name be `unmatchJobFromContact`?

Perhaps a clearer method name would be `removeJobFromContact`? Not very sure.

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    public static final String MESSAGE_UNMATCH_SUCCESS = "Unmatched Contact: %1$s with Job: %2$s";
+    public static final String MESSAGE_ALREADY_UNMATCHED = "Contact already unmatched with this job!";
+
+    private final Index contactIndex;
+    private final Index jobIndex;
+
+    /**
+     * @param contactIndex Index of the contact in the filtered person list to unmatch
+     * @param jobIndex Index of the job in the filtered contact list to unmatch
```
Comment (KengHian) [inline src/main/java/seedu/address/logic/commands/UnmatchCommand.java:44]: Punctuation behind each parameter description.

**Review by KengHian — APPROVED** [review]
Overall LGTM. I know we inherited some of the JavaDoc from AB3 (or from me opps), but I think we should change them according to the coding standards we have been provided. For more information, you can refer to my text in Telegram.

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java (hunk 1 above)
Comment (KengHian) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:89]: @Green-Tea-123 perhaps you could share your take on this?

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+        // TODO: Assume contact name is used to preserve uniqueness for now
+        final String contactIdentifier = contactToUnmatch.getIdentifier();
+        final String jobIdentifier = jobToUnmatch.getIdentifier();
+
+        // Check if the contact is already unmatched from the job
+        boolean hasContactMatchedJob = contactToUnmatch.hasMatched(jobIdentifier);
+        boolean hasJobMatchedContact = jobToUnmatch.hasMatched(contactIdentifier);
+
+        if (!hasContactMatchedJob && !hasJobMatchedContact) {
```
[the remark this answers is not in the document]
Comment (Green-Tea-123) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:82]: I agree that we should check if the contact and job are matched together before we try unmatching them.

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java (hunk 1 above)
Comment (Green-Tea-123) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:89]: I agree

**Review by Green-Tea-123 — COMMENTED** [review]
I think the main issue needed to be addressed is the check of whether the job and contact are actually matched together before they are unmatched.

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java (hunk 1 above)
Comment (KengHian) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:89]: @Green-Tea-123 thanks for the input! May I just check out of the 3 options I listed, which one do you think is the best?

Comment (Green-Tea-123) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:89]: I'd say the second as its meaning is clear and job and contact preserve their subject-object role.

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlekKwek18) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:82]: Good point, will add the logic to validate the contact and the job

File: src/main/java/seedu/address/logic/parser/UnmatchCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package seedu.address.logic.parser;
+
+import static seedu.address.logic.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+
+import seedu.address.commons.core.index.Index;
+import seedu.address.logic.commands.UnmatchCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+
+/**
+ * Parses input arguments and creates a UnmatchCommand object
```
Comment (KengHian) [inline src/main/java/seedu/address/logic/parser/UnmatchCommandParser.java:10]: Add punctuation after every sentence in a JavaDoc.

File: src/main/java/seedu/address/logic/parser/UnmatchCommandParser.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public class UnmatchCommandParser implements Parser<UnmatchCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the UnmatchCommand
+     * and returns a UnmatchCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
```
Comment (KengHian) [inline src/main/java/seedu/address/logic/parser/UnmatchCommandParser.java:16]: Same issue as above, for more information you can read my text in Telegram regarding the JavaDoc style.

**Review by KengHian — APPROVED** [review]
LGTM, gave some suggestions, just let me know if you want a further discussion on any of them.

**Review by Green-Tea-123 — APPROVED** [review]
LGTM now

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+     */
+    private void validateMatching(Person contactToUnmatch, Job jobToUnmatch) throws CommandException {
+        // Check if the contact has no jobs associated at all
+        if (contactToUnmatch.getMatch() == null || contactToUnmatch.getMatch().isEmpty()) {
+            throw new CommandException(String.format(MESSAGE_CONTACT_HAS_NO_JOBS, contactToUnmatch.getName(),
+                    jobToUnmatch.getName()));
+        }
+
+        // Check if the job has no contacts associated at all
+        if (jobToUnmatch.getMatches().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (AlekKwek18) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:129]: I agree. Will remove this check

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            throw new CommandException(String.format(MESSAGE_JOB_HAS_NO_CONTACTS, jobToUnmatch.getName(),
+                    contactToUnmatch.getName()));
+        }
+
+        // Check if the contact and job are not matched to each other
+        boolean hasContactMatchedJob = contactToUnmatch.hasMatched(jobToUnmatch.getIdentifier());
+        boolean hasJobMatchedContact = jobToUnmatch.hasMatched(contactToUnmatch.getIdentifier());
+
+        if (!hasContactMatchedJob || !hasJobMatchedContact) {
```
[the remark this answers is not in the document]
Comment (AlekKwek18) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:138]: Will fix this, thanks for pointing this out

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        return new Job(name, company, salary, description, requirements, matches);
+    }
+
+    /**
+     * Validates that the contact and job are matched to each other and that both
+     * entities have relevant matches.
+     */
+    private void validateMatching(Person contactToUnmatch, Job jobToUnmatch) throws CommandException {
+        // Check if the contact has no jobs associated at all
+        if (contactToUnmatch.getMatch() == null || contactToUnmatch.getMatch().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (AlekKwek18) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:123]: I agree. Will replace the check with !contactToUnmatch.isMatchPresent()

[the remark this answers is not in the document]
Comment (AlekKwek18) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:123]: To be addressed, determine if both conditions need to be checked

File: src/main/java/seedu/address/logic/commands/UnmatchCommand.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        this.contactIndex = contactIndex;
+        this.jobIndex = jobIndex;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+
+        List<Person> lastShownPeronList = model.getFilteredPersonList();
+        if (contactIndex.getZeroBased() >= lastShownPeronList.size()) {
```
Comment (Green-Tea-123) [inline src/main/java/seedu/address/logic/commands/UnmatchCommand.java:60]: I just realised that there is a misspelling lol

Comment (AlekKwek18) [reply src/main/java/seedu/address/logic/commands/UnmatchCommand.java:60]: Thanks for the catch!

**Pull request merged**